src/planning/src/planning/path_smoothing.cpp

```diff
@@ -0,0 +1,220 @@
+  int n = getPointAmount();
+  for (int i = 0; i < n; i++) {
+    std::cout << "(" << path[i] -> getX() << "," << path[i] -> getY() << "),";
+  }
+  std::cout << std::endl;
+}
+
+void PathSmoothing::fillPath(const std::string &path) {
+  RCLCPP_DEBUG(rclcpp::get_logger("rclcpp"), "START fillTrack");
+  std::string x, y, v;
+  std::ifstream trackFile = openReadFile(path);
+  while (trackFile >> x >> y >> v) {
+    float xValue = stof(x);
+    float yValue = stof(y);
+    float vValue = stof(v);
+    addPathPoint(xValue, yValue, vValue);
+  }
+  trackFile.close();
+  RCLCPP_DEBUG(rclcpp::get_logger("rclcpp"), "END fillTrack");
+}
+
+bool PathSmoothing::vector_direction(PathPoint *c1, PathPoint *c2, float prev_vx, float prev_vy) {
+  float vx = c2->getX() - c1->getX();
+  float vy = c2->getY() - c1->getY();
+  // Cos(angle) higher than 0 yhe vectors are pointing to the same half
+  return (vx * prev_vx + vy * prev_vy) > 0;
+}
+
+std::vector<PathPoint *> PathSmoothing::orderPath(std::vector<PathPoint *> *unord_path) {
+  // Order Pathpoint_array
+  // The algorithm works by iteratively selecting the nearest unvisited
+  // path_point to the current path_point and updating the traversal direction accordingly.
+  // This approach creates an ordered sequence that efficiently connects nearby
+  // path points.
+  std::vector<std::pair<PathPoint *, bool>> new_unord_path;
+  for (int i = 0; i < static_cast<int>(unord_path -> size()); i++)
+    new_unord_path.push_back(std::make_pair((*unord_path)[i], false));
+
+  // Values for first iteration
+  std::vector<PathPoint *> path;
+  PathPoint *path_point1 = new PathPoint(0, 0, 0);
+  float vx = 1;
+  float vy = 0;
+
+  for (int iter_number = 0; iter_number < static_cast<int>(unord_path->size());
+       iter_number++) {
+    float min_dist = MAXFLOAT;
+    int min_index = 0;
+
+    for (int i = 0; i < static_cast<int>(new_unord_path.size()); i++) {
+      PathPoint *path_point2 = new_unord_path[i].first;
+      // Check only if not visited
+      if (new_unord_path[i].second == false ||
+          (iter_number == 0 && vector_direction(path_point1, path_point2, vx, vy))) {
+        // first iteration we assure the direction is correct to avoid going
+        // backwards
+
+        float new_dist = path_point1->getDistanceTo(path_point2);
+        if (new_dist < min_dist) {
+          min_dist = new_dist;
+          min_index = i;
+        }
+      }
+    }
+    new_unord_path[min_index].second = true;  // mark as visited
+
+    // new visited path points is the reference for the next search. Arrays and path point
+    // updated
+    vx = new_unord_path[min_index].first->getX() - path_point1->getX();
+    vy = new_unord_path[min_index].first->getY() - path_point1->getY();
+    path_point1 = new_unord_path[min_index].first;
+
+    path.push_back(new_unord_path[min_index].first);  // add path points to ordered sequence
+  }
+  return path;
+}
+
+void PathSmoothing::validate(std::vector<PathPoint *> a_path) {
```

Comment:
I know it was called validate but I don't think the name is appropriate anymore as it changes the path drastically, it isn't really a validation

Reply:
✅